drivers: gpio: add interrupt support to renesas rzt2m gpio

wsipak opened the pull request:
Adds interrupt support to renesas rzt2m gpio

fkokosinski commented:
Hi @galak, @henrikbrixandersen, @mnkp. Sorry for pinging you guys, but could you take a look at this PR? Thanks

henrikbrixandersen reviewed (changes requested):
Please split this PR into multiple commits and avoid mixing driver SoC DTSI files where possible.

I'd add the binding changes first, modify the SoC DTSI, add the driver changes, then modify the boards.

jmichalski-ant committed:
dts: bindings: update renesas rzt2m gpio bindings to handle interrupts

Add irqs to rzt2m gpio bindings in order to add interrupt support to rzt2m
gpio driver and adds common gpio node to store interrupt config (irqs are
shared between ports)

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>

jmichalski-ant committed:
dts: arm: add nodes to support gpio interrupts in renesas rzt2m

Add common gpio node to pinctrl node (interrupts are shared between ports)
and syscon for interrupt edge detection register in order to support
interrupts in rzt2m gpio

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>

jmichalski-ant committed:
drivers: gpio: add interrupt support to renesas rzt2m gpio

Add interrupt support to renesas rzt2m gpio driver

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>

jmichalski-ant committed:
boards: add switch nodes to rzt2m starter kit

Add onboard switch nodes to rzt2m starter kit

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>

jmichalski-ant commented on drivers/gpio/gpio_rzt2m.c:
```diff
@@ -211,22 +251,206 @@ static int rzt2m_gpio_configure(const struct device *dev, gpio_pin_t pin, gpio_f
+{
+	const struct rzt2m_gpio_config *config = dev->config;
+
+	if (pin >= 8) {
```
I've used a define here

jmichalski-ant commented on drivers/gpio/gpio_rzt2m.c:
```diff
@@ -211,22 +251,206 @@ static int rzt2m_gpio_configure(const struct device *dev, gpio_pin_t pin, gpio_f
+	}
+
+	/* secure range - currently not supported*/
+	if (irq > 13) {
```
I've used a define here

jmichalski-ant commented on drivers/gpio/gpio_rzt2m.c:
```diff
@@ -211,22 +251,206 @@ static int rzt2m_gpio_configure(const struct device *dev, gpio_pin_t pin, gpio_f
+			   LISTIFY(DT_INST_PROP_LEN_OR(inst, irqs, 0), VALUE_2X, (,)))
+
+#define RZT2M_DECLARE_INIT(inst)                                                                   \
+	COND_CODE_0(inst, (static int rzt2m_gpio_init_##inst(const struct device *dev) \
```
The code related to the initialization of common parts of gpio ports (registration of interrupt handlers) that was previously invoked here has been rewritten since your comment. It was moved to the initialization of the one common instance for all ports

jmichalski-ant commented:
I've split it into separate comments as you suggested

henrikbrixandersen commented on drivers/gpio/gpio_rzt2m.c:
```diff
@@ -27,26 +39,55 @@
+	sys_slist_t cb;
+};
+
+struct rzt2m_gpio_common_config {};
```
Why this empty struct?

jmichalski-ant replied:
Instance of it is passed to the `DEVICE_DT_DEFINE` [here](https://github.com/zephyrproject-rtos/zephyr/pull/66485/commits/5b0cd8e684c79fe3941bfdd45edeb8251a03a807#diff-0861a4ed0138ef2fa644ca8f44d230d8429eb423811043d69e9ef5c216ced3efR433). While its not used (only data is) we have to pass something here because [documentation](https://docs.zephyrproject.org/apidoc/latest/group__device__model.html#gac49e26fbe91a14307d5ea08d41561dd1) doesn't allow it to be null

henrikbrixandersen replied:
I believe it is okay to use `NULL`?

jmichalski-ant replied:
I've removed this struct and passed `NULL` instead of it to `DEVICE_DT_DEFINE`

henrikbrixandersen commented on drivers/gpio/gpio_rzt2m.c:
```diff
@@ -27,26 +39,55 @@
+	struct rzt2m_gpio_irq_slot irq_registered_ports[IRQ_COUNT];
 };
 
+static struct rzt2m_gpio_common_config rzt2m_gpio_common_config = {};
+
+static struct rzt2m_gpio_common_data rzt2m_gpio_common_data = {};
```
No need for initializers here?

jmichalski-ant replied:
They are not needed and I've removed them. At the same time I had to rename those because the same name as the type name without initializer was causing ci compliance errors

henrikbrixandersen commented on drivers/gpio/gpio_rzt2m.c:
```diff
@@ -211,22 +268,193 @@ static int rzt2m_gpio_configure(const struct device *dev, gpio_pin_t pin, gpio_f
+
+	static uint8_t n[IRQ_COUNT];
+
+	for (int i = 0; i < sizeof(n); i++) {
```
```suggestion
	for (int i = 0; i < ARRAY_SIZE(n); i++) {
```

jmichalski-ant replied:
I've applied your suggestion

wsipak commented:
@henrikbrixandersen I believe your comments have been addressed. Could you take a look at this again? Thanks a lot!

fkokosinski commented:
Hey @fabiobaltieri! Is there something that's currently blocking this PR from being merged and requires action from our side? Thanks :)

fabiobaltieri commented:
@fkokosinski hey we are holding back at merging PRs touching files in boards/ waiting for the hwmv2 collab branch merge, apologies for the delay, that should tentatively resolved by end of week

nordicjm reviewed (dismissed):
Needs converting to hwmv2

jmichalski-ant commented:
I've updated it to use `boards/renesas` instead of `boards/arm`. Hope its compliant with hwmv2 now

fkokosinski commented:
@henrikbrixandersen @tgorochowik could you PTAL? Thanks!

fkokosinski commented:
@henrikbrixandersen sorry for pinging you again, but it looks like your comments are addressed. Could you PTAL? Thanks!

wsipak commented:
I've removed my assignment as I'm the author.
